algebird-core/src/main/scala/com/twitter/algebird/Window.scala

```diff
@@ -0,0 +1,120 @@
+/*
+Copyright 2018 Stripe
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package com.twitter.algebird
+
+import scala.collection.immutable.Queue
+
+/**
+ *
+ * Convenience case class defined with a monoid for aggregating elements over
+ * a finite window.
+ *
+ * @param total Known running total of `T`
+ * @param items queue of known trailing elements.
+ */
+
+case class Window[T](total: T, items: Queue[T]) {
+  def size = this.items.size
+}
+
+object Window {
+  def apply[T](v: T): Window[T] = Window[T](v, Queue[T](v))
+}
+
+/**
+ * Provides a natural monoid for combining windows truncated to some window size.
+ *
+ * @param windowSize
+ */
+
+
+case class WindowMonoid[T](
+    windowSize: Int
+)(implicit p: Priority[Group[T], Monoid[T]])
+    extends Monoid[Window[T]] {
+  val zero =
+    p match {
+      case Priority.Preferred(g) => Window(g.zero)
+      case Priority.Fallback(m)  => Window(m.zero)
+    }
+
+  def plus(a: Window[T], b: Window[T]): Window[T] =
+    p match {
+      case Priority.Preferred(g) => plusG(a, b)(g)
+      case Priority.Fallback(m)  => plusM(a, b)(m)
+    }
+
```

Comment:
I think you can optimize `sumOption` somewhat here: you can take blocks of data and ignore the left if the right is full. PS: scalding uses this if it exists, so overridding sumOption on Semigroup (or any subclass) can be a very nice performance win on map/reduce.

```scala
def sumOption(ws: TraversableOnce[Window[T]]): Option[Window[T]] =
  if (ws.isEmpty) None
  else {
    val it = ws.toIterator
    var size = 0
    var queue = Queue.empty[Window[T]]
    while (it.hasNext) {
      val n = it.next
      queue = queue :+ n
      size += n.size
      val tailSize = size - queue.head.size
      if (tailSize >= windowSize) {
        queue = queue.tail
        size = tailSize
      }
    }
    // now we only have to merge the queue:
    queue.tail.foldLeft(queue.head)(plus(_, _))
  }
```
I think this can be a pretty big win.

Reply:
@johnynek: Overwrote sumOption below.